lib/req/steps.ex

```diff
@@ -2059,6 +2079,220 @@ defmodule Req.Steps do
+            request
+            |> Req.Request.put_header("authorization", auth_header_value)
+            |> Req.Request.put_private(:http_digest_attempted, true)
+            |> Req.Request.run_request()
+
+          {:error, _reason} ->
+            {request, response}
+        end
+
+      {:error, :no_digest_challenge} ->
+        {request, response}
+    end
+  end
+
+  defp get_challenge_header(%Req.Response{status: 401} = response) do
+    case Req.Response.get_header(response, "www-authenticate") do
+      [auth_header | _] ->
+        {:ok, auth_header}
+
+      _ ->
+        {:error, :no_digest_challenge}
+    end
+  end
+
+  defp parse_challenge(header) do
+    header
+    |> String.trim_leading("Digest ")
+    |> String.split(",")
+    |> Enum.map(&String.trim/1)
+    |> Map.new(&parse_header_part/1)
+  end
+
+  defp parse_header_part(part) do
+    case String.split(part, "=", parts: 2) do
+      [key, value] ->
+        {String.trim(key), unquote_value(value)}
+
+      [key] ->
+        {String.trim(key), ""}
+    end
+  end
+
+  defp count_to_nc(count) do
+    String.pad_leading(Integer.to_string(count, 16), 8, "0")
+  end
+
+  defp generate_auth_header(challenge, username, password, method, uri, count) do
+    algorithm = Map.get(challenge, "algorithm", "MD5")
+    realm = Map.get(challenge, "realm", "")
+    nonce = Map.get(challenge, "nonce", "")
+    opaque = Map.get(challenge, "opaque")
+    qop = Map.get(challenge, "qop")
+    cnonce = generate_cnonce()
+    nc = count_to_nc(count)
+
+    with {:ok, hash_func} <- hash_function(algorithm) do
+      ha1 = hash_func.("#{username}:#{realm}:#{password}")
+
+      ha1 =
+        if algorithm |> String.upcase() |> String.ends_with?("-SESS") do
+          hash_func.("#{ha1}:#{nonce}:#{cnonce}")
+        else
+          ha1
+        end
+
+      method_str = method |> to_string() |> String.upcase()
+      ha2 = hash_func.("#{method_str}:#{uri}")
+
+      response =
+        if qop in ["auth", "auth-int"] do
+          hash_func.("#{ha1}:#{nonce}:#{nc}:#{cnonce}:#{qop}:#{ha2}")
+        else
+          hash_func.("#{ha1}:#{nonce}:#{ha2}")
+        end
+
+      build_auth_header(
+        username: username,
+        realm: realm,
+        nonce: nonce,
+        uri: uri,
+        response: response,
+        qop: qop,
+        nc: nc,
+        cnonce: cnonce,
+        opaque: opaque,
+        algorithm: algorithm
+      )
+    end
+  end
+
+  defp md5_hash(data) do
+    :md5
+    |> :crypto.hash(data)
+    |> Base.encode16(case: :lower)
+  end
+
+  defp sha256_hash(data) do
+    :sha256
+    |> :crypto.hash(data)
+    |> Base.encode16(case: :lower)
+  end
+
+  defp generate_cnonce do
+    16
+    |> :crypto.strong_rand_bytes()
+    |> Base.encode16(case: :lower)
+  end
+
+  def hash_function(algorithm) do
+    case String.upcase(algorithm) do
+      "MD5" <> _ -> {:ok, &md5_hash/1}
+      "SHA-256" <> _ -> {:ok, &sha256_hash/1}
+      # TODO: Add support for SHA-512-256
```

Comment:
do you intend to implement this in this PR? If not, which is totally fine, let's remove this todo and wait until it comes up.